### PR TITLE
[74354] Add a simple login XWiki button

### PR DESCRIPTION
**Pull request by shiroginne**
# Ticket
[74354](https://community.openproject.org/projects/stream-xwiki-integration/work_packages/74354/)

# What are you trying to accomplish?
Add an XWiki login button to the wiki tab. Update the admin's connect button to have the same icon. 
Fixed an issue with the white spade in the xwiki delete modal window. 

## Screenshots
Screenshot 2026-04-28 at 10 38 33

Screenshot 2026-04-28 at 15 45 55

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb [hunk 1]
```diff
@@ -34,7 +34,11 @@ See COPYRIGHT and LICENSE files for more details.
        message.with_heading(tag: :h2) { t(".title") }
      end %>
   <% dialog.with_additional_details do %>
-    <%= t(".warning_html", wiki_provider: content_tag(:strong, @wiki_provider.name)) %>
+    <% render(Primer::OpenProject::FlexLayout.new) do |flex| %>
```
Comment (shiroginne) [inline modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: We need the flex layout to render correct spaces = no flex layout, no space rendered for <strong> text 😱

Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: Not sure which space you are referring to. Whitespace in the middle of text? Padding?

I am curious. If it looks like a bug to you, we should tell the design team.

File: modules/wikis/app/components/wikis/oauth_login_component.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Wikis
+  class OAuthLoginComponent < BaseComponent
+    def login_url
+      oauth_clients_ensure_connection_url(
+        oauth_client_id: provider.oauth_client.client_id,
+        storage_id: provider.id,
```
Comment (NobodysNightmare) [inline modules/wikis/app/components/wikis/oauth_login_component.rb:46]: 🟡 I just noticed this now. Is there any chance (maybe even in a followup PR), that we can rename this named parameter to `integration_id`?

File: modules/wikis/app/components/wikis/oauth_login_component.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        destination_url: work_package_url(@work_package, tab: :wikis)
```
Comment (NobodysNightmare) [inline modules/wikis/app/components/wikis/oauth_login_component.rb]: It feels like a separation of concerns issue to me that this component knows where it has to redirect to, because effectively it means that the component knows the path on which it's rendered.

For example on the inbox view this will already lead to a slightly wrong return url.

What do you think of accepting this as a initializer argument?

```ruby
def initialize(return_url:, **)
  super(**)
  @return_url = return_url
end
```

This would allow the embedding component to decide where to redirect to.

...

Second thought: This would probably not be enough, would it? Because the embedding component also has no clue where it lives... meh...

File: modules/wikis/app/components/wikis/relation_page_links_component.rb [hunk 4]
```diff
@@ -29,21 +29,17 @@
...
-      super(model, **)
-    end
-
+  class RelationPageLinksComponent < BaseComponent
     def page_link_infos
       @page_link_infos ||= page_link_service.relation_page_link_infos_for(provider:, linkable: @work_package)
     end
 
+    def user_connected?
+      return true if provider.oauth_client.blank?
```
Comment (NobodysNightmare) [inline modules/wikis/app/components/wikis/relation_page_links_component.rb]: 🔴 This will raise an error for providers that don't have an `oauth_client` method (e.g. the `InternalProvider`).

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb [hunk 5]
```diff
@@ -52,9 +52,13 @@ See COPYRIGHT and LICENSE files for more details.
 
     if page_link_infos.empty?
       box.with_row do
-        render(Primer::Beta::Blankslate.new(border: false)) do |blankslate|
-          blankslate.with_heading(tag: :h2).with_content(t(".empty_heading"))
-          blankslate.with_description { t(".empty_text") }
+        if user_connected?
```
Comment (NobodysNightmare) [inline modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: 🟡 It looks weird to me that we only do the `user_connected?` check if there are no `page_link_infos`. This includes a lot of assumptions about the behaviour of the `page_link_service`, mostly that unauthenticated requests cause the service to return zero results (instead of an error or instead of a result with some error entities).

Shouldn't we first check whether there is a connection and only if there is one try to fetch and act on page link infos?

Or do we explicitly want to not show a login button, if fetching _something_ worked? Especially for XWiki it's thinkable that some content could be fetched anonymously...

File: modules/wikis/config/locales/en.yml [hunk 6]
```diff
@@ -40,6 +40,8 @@ en:
       empty_heading: No related pages
       empty_text: Manually add links to other related wiki pages.
     oauth_login_component:
+      heading: Connect your %{provider} account
+      description: Log in to %{provider} to view and manage related wiki pages.
```
Comment (NobodysNightmare) [inline modules/wikis/config/locales/en.yml]: This is the part that we wanted to discuss on the PR, right?

Let me add a proposal (CC @Kharonus)


```yaml
      heading: Not connected to %{provider}
      description: Log in to %{provider} to view and manage related wiki pages from this OpenProject instance.
```

File: modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  let(:oauth_client) { create(:oauth_client, integration: provider) }
+
+  let(:page_link_service) { instance_double(Wikis::PageLinkService, relation_page_link_infos_for: []) }
+
+  before do
+    login_as(user)
+    allow(Wikis::PageLinkService).to receive(:new).and_return(page_link_service)
+  end
+
+  context "when the provider has no oauth client configured" do
```
Comment (NobodysNightmare) [inline modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb:49]: Since I talked about this failing for the internal provider:

```ruby
context "when the provider does not support OAuth" do
  let(:provider) { create(:internal_wiki_provider) }

  # it ...
end
```

File: modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+require "spec_helper"
+require_module_spec_helper
+
+RSpec.describe Wikis::RelationPageLinksComponent, type: :component do
+  let(:user) { create(:user) }
+  let(:work_package) { build_stubbed(:work_package) }
+  let(:provider) { create(:xwiki_provider) }
+  let(:oauth_client) { create(:oauth_client, integration: provider) }
+
+  let(:page_link_service) { instance_double(Wikis::PageLinkService, relation_page_link_infos_for: []) }
```
Comment (NobodysNightmare) [inline modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb:40]: 🟢 I see that this should have happened in a previous PR already, but it would be nice to at least have one simple spec covering the case where the component is rendering what it's supposed to render under "normal operation". Right now this is purely an edge case spec, that tests the components ability to not render any page links.

**Review by NobodysNightmare — CHANGES_REQUESTED** [review]
Poor `InternalProvider` has to suffer again :(

**Comment by shiroginne** [comment]
> Poor `InternalProvider` has to suffer again :(

The `InternalProvider` Strikes Back ☬

I'll create `InternalProvider` in my local DB to prevent this type of errors, thanks for pointing out! (I guess, we need to seed it to QA env as well, so QA would know about it)

**Comment by NobodysNightmare** [comment]
> (I guess, we need to seed it to QA env as well, so QA would know about it)

Let's not do anything special on QA. The plan is to _eventually_ seed this provider on **every** OP instance. It should always exist and can then be disabled by users, if needed.

File: modules/wikis/app/components/wikis/oauth_login_component.rb (hunk 3 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/oauth_login_component.rb]: This would decouple the OAuth component from the URL hardcoded 👌🏻 but it's true, the component above - relation link page component then would have to have this URL hardcoded. I think it's okay. Other options could be `request.referer` for example. We could get the current URL to pass to the component, but I'd rather have it hardcoded 🤔

File: modules/wikis/app/components/wikis/oauth_login_component.rb (hunk 2 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/oauth_login_component.rb:46]: I'd say in the next PR

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb (hunk 1 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: Screenshot 2026-04-28 at 15 25 10

this is the one: there is a space in the text, but no in on the screen and the issue is "strong" tag. 

I should bring it to the design team.

File: modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb (hunk 8 above)
Comment (shiroginne) [reply modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb:40]: I'd say it would be ideal for https://community.openproject.org/projects/stream-xwiki-integration/work_packages/74542/activity issue. Right now the view is hardcoded 😅 

```
          render(Primer::Beta::Blankslate.new(border: false)) do |blankslate|
            blankslate.with_heading(tag: :h2).with_content(t(".empty_heading"))
            blankslate.with_description { t(".empty_text") }
          end

```

File: modules/wikis/app/components/wikis/relation_page_links_component.rb (hunk 4 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/relation_page_links_component.rb]: Ok, this one is a tricky one, as it's not the first occurrence with the different behaviour, I could think about to add `internal?` method to `Wiki::Provider` and then override it to `true` for the `InternalProvider` so we would know which one is rendering right now 🤔

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb (hunk 5 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: If I remember correctly, we decided to show a log-in button if the user is not logged in, even if there are pages which could be fetched “anonymously", right @Kharonus? 

If it's not the case, I think this part of the view should be updated in the issue where we actually fetching something.

File: modules/wikis/app/components/wikis/oauth_login_component.rb (hunk 3 above)
Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/oauth_login_component.rb]: `request.referer` is not guaranteed to be filled, that's nothing I'd use.

In the end the value should be "hardcoded" somewhere, but by the place that can surely know the correct route.

But yeah, it's okay, we can leave it as is right now.

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb (hunk 1 above)
Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: Yes definitely. The HTML seems to be `some word <strong>XWiki</strong>.`, so the HTML is correct and includes whitespace before the strong tag. If the surrounding HTML causes this space to be eaten, this sounds like a bug to me.

CC @HDinger

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb [hunk 9]
```diff
@@ -34,7 +34,11 @@ See COPYRIGHT and LICENSE files for more details.
...
+      <% flex.with_row do %>
+        <%= t(".warning_html", wiki_provider: content_tag(:strong, @wiki_provider.name)) %>
```
Comment (NobodysNightmare) [inline modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: While researching for the eaten space above, I noticed that the `warning_html` translation is defined twice, but only used once... Can you delete the superfluous one?

File: modules/wikis/app/components/wikis/relation_page_links_component.rb (hunk 4 above)
Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/relation_page_links_component.rb]: I'd do _something_ else.

Using `internal?` has some strong assumptions on the internal provider being special. Then you could effectively also do `is_a?(InternalProvider)`, when Ruby idiomatically is about duck typing.

I don't think that checking `respond_to?(:some_method)` is a strong code smell, maybe others have a stronger opinion on this, but that's probably what I would have done.

Alternatively we'd add something around the authentication method, because the check below is kinda tied to the provider using a special kind of auth method and would fail for others.

Or we delegate the `user_connected?` check to the provider or one of its queries, this would allow the `InternalProvider` to effectively accept all users, while other providers could decide based on their authentication method (remember that we initially planned to support at least two ways of authentication, this would also fail with the approach below).

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb (hunk 1 above)
Comment (HDinger) [reply modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: Okay, let's split that problem a bit: I think the `additional_details` slot is the wrong one. You should be using the message description:

```
 <% dialog.with_confirmation_message do |message|
         message.with_heading(tag: :h2) { t(".title") }
         message.with_description_content(
            t(".warning_html", wiki_provider: content_tag(:string, @wiki_provider.name))
         )
       end %>
```

That will properly render as expected and is semantically the correct slot

--- 

Fyi: Let's quickly look at what happened in the additional_details slot: The `additional_details` slot is designed to render **any** kind of additional stuff.. This can be text, images, banners, etc. So it is a shallow container that takes care of alignmnet of its children with a flex box logic. 
The original implementation misused that slightly by not passing an HTML element, but the string directly. Since this string itself contains an HTML tag, the browser splits the string into three elements. That then looks like this:

Image

The container cannot know that this is a text only block and aligns with the standard flex machanism.

So if you ever encounter that problem again, you don't need a flex layout, but create proper HTML by wrapping the text into a `p`. 

``` 
<% dialog.with_additional_details do %>
   <p>
     <%= t(".warning_html", wiki_provider: content_tag(:string, @wiki_provider.name)) %>
   </p>
<% end %>
```

---

Tl;dr: Use the `message.with_description` slot explained at the beginning and you are done 👍 

@NobodysNightmare @shiroginne

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb [hunk 10]
```diff
@@ -34,7 +34,11 @@ See COPYRIGHT and LICENSE files for more details.
        message.with_heading(tag: :h2) { t(".title") }
```
Comment (HDinger) [inline modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: The heading of a danger dialog should ideally be formulated as a question, something like "Delete wiki provider?". Thereby, it is important to mention, that the heading of the DangerDialog should remain in a neutral tone (e.g. "Delete wiki provider") since this is read as the name of the dialog to screen reader users. Currently both use the same string. This pattern is also documented [here](https://qa.openproject-edge.com/lookbook/pages/components/danger_dialog)

I know that this is not part of your PR, but maybe you could update it nevertheless?

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb [hunk 11]
```diff
@@ -52,9 +52,18 @@ See COPYRIGHT and LICENSE files for more details.
...
+          render(Primer::Beta::Blankslate.new(border: false)) do |blankslate|
+            blankslate.with_heading(tag: :h2).with_content(t(".empty_heading"))
+            blankslate.with_description { t(".empty_text") }
+          end
+        else
+          render(Wikis::OAuthLoginComponent.new(
+            provider,
+            return_url: work_package_url(@work_package, tab: :wikis),
+            work_package: @work_package
```
Comment (Kharonus) [inline modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: 🔴 what is that work package for here? There shouldn't be any work package context needed for the login, right? 

I mean, you need the return url, but this is already constructed in the line above.

File: modules/wikis/app/components/wikis/base_component.rb
Comment (Kharonus) [inline modules/wikis/app/components/wikis/base_component.rb:1]: 🟡 Is really a base component needed? in this namespace it kind of gives the wrong vibes, as it adds a work package context - plainly. But in this namespace their COULD be other components that never need that context.

I'd rather leave the alias mapping and the context in the `intialize` method to the components itself, because they differ.

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb (hunk 5 above)
Comment (Kharonus) [reply modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: > If I remember correctly, we decided to show a log-in button if the user is not logged in, even if there are pages which could be fetched “anonymously", right [@Kharonus](https://github.com/Kharonus)?

Yes, but this means, the component should show the login over any logic of content, right? we don't care - yet - if there were any page links fetched.

File: modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb (hunk 1 above)
Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: Thanks for the detailed explanation.

Comment (shiroginne) [reply modules/wikis/app/components/wikis/admin/destroy_confirmation_dialog_component.html.erb]: Thanks @HDinger 👍🏻

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb (hunk 11 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: Good one, thanks! Leftover from the times when a url was a part of the `OAuthLoginComponent`. Removed 🆗

File: modules/wikis/app/components/wikis/base_component.rb
Comment (shiroginne) [reply modules/wikis/app/components/wikis/base_component.rb:1]: It's not needed any more. Was useful before the refactoring started, now it's obsolete ... thou I like the `BaseComponent` in general to store all "common" logic for the components within some domain, but for now I'll remove it 👌🏻

File: modules/wikis/app/components/wikis/relation_page_links_component.rb (hunk 4 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/relation_page_links_component.rb]: Something like this? https://github.com/opf/openproject/pull/22924/commits/6917c73b63e957befd88d1928b963e64f728be12

File: modules/wikis/app/components/wikis/oauth_login_component.rb (hunk 2 above)
Comment (shiroginne) [reply modules/wikis/app/components/wikis/oauth_login_component.rb:46]: Here it is → https://github.com/opf/openproject/pull/23049

File: modules/wikis/app/components/wikis/relation_page_links_component.rb (hunk 4 above)
Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/relation_page_links_component.rb]: yes

File: modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb (hunk 8 above)
Comment (NobodysNightmare) [reply modules/wikis/spec/components/wikis/relation_page_links_component_spec.rb:40]: Below that:

```
    page_link_infos.each do |info|
      box.with_row { render(Wikis::PageLinkComponent.new(info, actions: [:remove])) }
    end
```

It is actually capable of rendering stuff and I saw that already ^^

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb (hunk 5 above)
Comment (NobodysNightmare) [reply modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: So does that mean:

```ruby
if !user_connected?
  # render login blank-slate
elsif page_link_infos.empty?
 # render no results blank slate
end

page_link_infos.each do |page_info|
  # render page link row
end
```

Or am I misunderstanding what we need?

File: modules/wikis/config/locales/en.yml (hunk 6 above)
Comment (NobodysNightmare) [reply modules/wikis/config/locales/en.yml]: Going to resolve this, since you took my proposal and nobody else commented that they hated it.

**Review by NobodysNightmare — COMMENTED** [review]
This mostly LGTM already, though not all conversations have concluded yet. I resolved what I could resolve (also as a proxy for our guest reviewer Henriette). What's open is the correct timing to render the button (with/without content). I don't have a strong opinion here, but tried to paraphrase what @Kharonus wrote in code, to see whether I understood it correctly.

File: modules/wikis/app/components/wikis/relation_page_links_component.html.erb (hunk 5 above)
Comment (Kharonus) [reply modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: No, you are not misunderstanding it. I think, this is the correct structure.

Comment (shiroginne) [reply modules/wikis/app/components/wikis/relation_page_links_component.html.erb]: Updated the code to match the description above 🔝
